### ISSUE TITLE
video_inference.py took too long 

### ISSUE THREAD

**Issue by mzalaki00** [xg416/TMT#5]
Hi, so thanks for your models
i run:
 video_inference.py  --input_path ./TMT/video/my_video.avi --out_path my_video_out.mp4 
--model_path ../TMT_dynamic.pth --save_video

My video is a short 2-second turbulence video. However, when I ran the above script, it took about 2 hours to complete without any issues. I'm wondering if I am using the TMT_dynamic.pth model or video_inference.py correctly to restore video or if the configuration parameters such as --patch_size and --temp_patch are affecting speed??

**Comment by xg416** [comment]
Hi, could you provide more information? What is the resolution and FPS of your video? What GPU are you using?
It is a weird speed. Could you check how long it takes to process a 240x240 patch on your machine? 

Note: with 2080Ti, our testing speed is 1 FPS without temporal overlapping and 0.5 FPS with temporal overlapping for 960x540 videos. If your video is 960x540, 60 frames, it should take 1~2 minutes to process.

**Comment by mzalaki00** [comment]
@xg416 
Hi again, my video is 640*480, 25fps, and duration is about 3seconds, gpu is rtx2060 , and packages
![Screenshot (547)](https://github.com/xg416/TMT/assets/91794577/521341df-1737-47e7-9995-461ef80d21c2)

how to to check a 240x240 patch on my machine?

**Comment by xg416** [comment]
Since the default setting of patch_size in video_inference.py is 240, you can just check how long it takes to execute line 38 in this script.

**Comment by mzalaki00** [comment]
@xg416 cause i use default settings, probably this long running time dedicated to this line 38.
i feel the code doesnt use gpu, even though my torch is enabled with cuda and in your code exist  torch.cuda.set_device(0)
i dont see any Sensible load on my gpu!!

**Comment by xg416** [comment]
The GPU is not loaded explains the slow speed. However, I have never seen a similar case before. I think you can try to make sure that your GPU can load the model and data.

**Comment by mzalaki00** [comment]
@xg416 i checked these cases to assure it use gpu and everything appears to be working correctly..
import torch
print(torch.cuda.is_available())  --> ans - True

net = net.cuda()
checkpoint = torch.load(args.model_path, map_location='cuda:0')

for name, param in net.named_parameters():
    assert param.device.type == 'cuda', f"Parameter {name} is not on the GPU!"

 input_ = torch.stack(inp_imgs, dim=1).unsqueeze(0).cuda()

the code to run: video_inference.py --patch_size 240  --input_path my_video.avi --out_path my_video.mp4 --model_path ./TMT_dynamic.pth --save_video

any idea to make it run faster or at least your 1fps??

**Comment by xg416** [comment]
How long it takes to run line 38?
`output_ = model(input_)`

**Comment by mzalaki00** [comment]
@xg416 for each iteration of     
for hi in hpos:
        for wi in wpos:
            input_ = input_blk[..., hi:hi+patch_size, wi:wi+patch_size]
            start_time = time.time()
            output_ = model(input_)
            print(f'time consumption {hi} for model(input_) line38 {time.time() - start_time} s')

result says:
time consumption 0 for model(input_) line38 6.53023886680603 s
time consumption 0 for model(input_) line38 102.18913626670837 s
time consumption 0 for model(input_) line38 102.11125636100769 s
time consumption 120 for model(input_) line38 101.75633430480957 s
time consumption 120 for model(input_) line38 110.78848123550415 s
time consumption 120 for model(input_) line38 108.7780487537384 s
.
.
.

**Comment by xg416** [comment]
Hi, here's what I get:
time consumption 0 for model(input_) line38 1.2594954833984375 s
time consumption 0 for model(input_) line38 1.162944580078125 s
time consumption 0 for model(input_) line38 1.1642843017578124 s
time consumption 120 for model(input_) line38 1.1632100830078125 s
time consumption 120 for model(input_) line38 1.1659898681640626 s
time consumption 120 for model(input_) line38 1.1644227294921874 s

I think there's some problem with your GPU, either because of the intrinsic issues or compatibility issues.

**Comment by mzalaki00** [comment]
ok thanks i test it again
 but i wonder why you use cuda().train() for inference !?

**Comment by xg416** [comment]
Hi, it should be eval(). In the previous version, I set no_grad_ manually for all parameters. Please update the code to the new version

**Comment by 1vanKa** [comment]
Hello, please, could you explain how did you get so fast inference on 2080Ti? I have 3090 and single model call took 4.2s with `patch_size = 240`.
I'm on Ubuntu 22.04, Intel Core i7-13700F, RTX 3090. nvidia-smi tells that it is 100% loaded.
```
Mon Feb 19 12:12:19 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.23.08              Driver Version: 545.23.08    CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3090        On  | 00000000:01:00.0  On |                  N/A |
| 51%   72C    P2             300W / 350W |  11210MiB / 24576MiB |    100%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|

...

|    0   N/A  N/A    351243      C   .../venvs/segmentation_venv/bin/python     8020MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by xg416** [comment]
For the speed, are you referring to the same scenarios as the following?
> time consumption 0 for model(input_) line38 1.2594954833984375 s
> time consumption 0 for model(input_) line38 1.162944580078125 s
> time consumption 0 for model(input_) line38 1.1642843017578124 s
> time consumption 120 for model(input_) line38 1.1632100830078125 s
> time consumption 120 for model(input_) line38 1.1659898681640626 s
> time consumption 120 for model(input_) line38 1.1644227294921874 s

I'm using ubuntu 20.04, Intel® Core™ i7-6700K CPU @ 4.00GHz.
NVIDIA-SMI 470.223.02   Driver Version: 470.223.02   CUDA Version: 11.4.
I have no idea why your machine could be slower, I have another 3080 machine, it is actually faster.

**Comment by 1vanKa** [comment]
For measuring speed i'm using small wrapper:
```
def named_timeit(name: str):
    def timeit(func):
        @wraps(func)
        def timeit_wrapper(*args, **kwargs):
            start_time = time.perf_counter()
            result = func(*args, **kwargs)
            end_time = time.perf_counter()
            total_time = end_time - start_time
            print(f'Function {name} Took {total_time:.4f} seconds')
            return result
        return timeit_wrapper
    return timeit
```
and measure like this:
```
    net = TMT_MS(num_blocks=[2,3,3,4], 
                    heads=[1,2,4,8], 
                    num_refinement_blocks=2, 
                    warp_mode='none', 
                    n_frames=args.temp_patch, 
                    att_type='shuffle').cuda().eval()
    # load
    checkpoint = torch.load(args.model_path)
    net.load_state_dict(checkpoint['state_dict'] if 'state_dict' in checkpoint.keys() else checkpoint)
    for name, param in net.named_parameters():
        param.requires_grad = False
    
    net = named_timeit("TMT")(net)
```
After this every net call prints its execution time.
Maybe you are using TensorRT with Torch?

**Comment by xg416** [comment]
Please align with our testing pipeline posted in the discussion above.

**Comment by 1vanKa** [comment]
It's same, but ok. Output:
```
0/1146
time consumption 0 for model(input_) line38 4.445489406585693 s
time consumption 0 for model(input_) line38 0.6641995906829834 s
time consumption 0 for model(input_) line38 4.232755661010742 s
time consumption 0 for model(input_) line38 4.23260235786438 s
time consumption 0 for model(input_) line38 4.297187566757202 s
time consumption 176 for model(input_) line38 4.235290765762329 s
time consumption 176 for model(input_) line38 4.209496974945068 s
time consumption 176 for model(input_) line38 4.26820707321167 s
time consumption 176 for model(input_) line38 4.232264757156372 s
time consumption 176 for model(input_) line38 4.2713398933410645 s
```

**Comment by xg416** [comment]
Can you test it on other machines? I will also try other machines.
I'll keep this issue open to see if anyone else has the same problem.

**Comment by 1vanKa** [comment]
On Ubuntu 20.04, RTX 3060, Driver Version: 515.65.01, CUDA Version: 11.7 every model call is ~6.8s.

**Comment by xg416** [comment]
Hi, I tried to use another machine with RTX 3080, Driver Version: 470.223.02, CUDA Version: 11.4, and 11th Gen Intel® Core™ i7-11700KF @ 3.60GHz × 16 processor. Because of the memory limit, I can only use patch size 224x224. The output for the above testing is:

```
0/56
time consumption 0 for model(input_) line38 0.5099799633026123 s
time consumption 0 for model(input_) line38 0.5218708515167236 s
time consumption 0 for model(input_) line38 0.8331954479217529 s
time consumption 0 for model(input_) line38 0.8359670639038086 s
time consumption 186 for model(input_) line38 0.8341319561004639 s
time consumption 186 for model(input_) line38 0.8373007774353027 s
time consumption 186 for model(input_) line38 0.8350830078125 s
time consumption 186 for model(input_) line38 0.8370840549468994 s
```

**Comment by xg416** [comment]
Here is what I got running `pip freeze > requirements_current.txt` on my RTX 3080 machine:

```
certifi==2023.7.22
cffi==1.15.1
charset-normalizer==3.3.0
click @ file:///home/conda/feedstock_root/build_artifacts/click_1666798198223/work
cloudpickle @ file:///home/conda/feedstock_root/build_artifacts/cloudpickle_1674202310934/work
comm==0.1.3
contourpy==1.0.7
cycler==0.11.0
cytoolz @ file:///home/conda/feedstock_root/build_artifacts/cytoolz_1666829688252/work
daal4py==2023.1.1
dask @ file:///home/conda/feedstock_root/build_artifacts/dask-core_1677271439707/work
debugpy==1.6.7
decorator==5.1.1
defusedxml==0.7.1
einops @ file:///home/conda/feedstock_root/build_artifacts/einops_1670600230829/work
executing==1.2.0
fastjsonschema==2.16.3
flit_core @ file:///opt/conda/conda-bld/flit-core_1644941570762/work/source/flit_core
fonttools==4.38.0
fqdn==1.5.1
fsspec @ file:///home/conda/feedstock_root/build_artifacts/fsspec_1674184942191/work
fvcore==0.1.5.post20221221
idna==3.4
image-quality @ file:///home/conda/feedstock_root/build_artifacts/image-quality_1621954419556/work
imagecodecs @ file:///opt/conda/conda-bld/imagecodecs_1664561985385/work
imageio @ file:///home/conda/feedstock_root/build_artifacts/imageio_1677507485990/work
importlib-metadata==6.1.0
iopath==0.1.10
ipykernel==6.22.0
ipython==8.12.0
ipython-genutils==0.2.0
isoduration==20.11.0
jedi==0.18.2
Jinja2==3.1.2
joblib @ file:///croot/joblib_1685113087166/work
jsonpointer==2.3
jsonschema==4.17.3
jupyter-events==0.6.3
jupyter_client==8.1.0
jupyter_core==5.3.0
jupyter_server==2.5.0
jupyter_server_terminals==0.4.4
jupyterlab-night==0.4.6
jupyterlab-pygments==0.2.2
kiwisolver==1.4.4
libsvm @ file:///home/conda/feedstock_root/build_artifacts/libsvm-python_1667111001705/work
locket @ file:///home/conda/feedstock_root/build_artifacts/locket_1650660393415/work
lpips==0.1.4
MarkupSafe==2.1.2
matplotlib==3.6.3
matplotlib-inline==0.1.6
mistune==2.0.5
mkl-fft==1.3.1
mkl-random @ file:///tmp/build/80754af9/mkl_random_1626186066731/work
mkl-service==2.4.0
nbclassic==0.5.5
nbclient==0.7.3
nbconvert==7.3.0
nbformat==5.8.0
nest-asyncio==1.5.6
networkx @ file:///home/conda/feedstock_root/build_artifacts/networkx_1673151334029/work
notebook==6.5.4
notebook_shim==0.2.2
numpy==1.22.4
opencv-python==4.7.0.68
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1673482170163/work
panda==0.3.1
pandas==1.5.3
pandocfilters==1.5.0
parso==0.8.3
partd @ file:///home/conda/feedstock_root/build_artifacts/partd_1660316728562/work
patsy @ file:///home/conda/feedstock_root/build_artifacts/patsy_1665356157073/work
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.3.0
platformdirs==3.2.0
portalocker==2.8.2
prometheus-client==0.16.0
prompt-toolkit==3.0.38
psutil==5.9.4
ptyprocess==0.7.0
pure-eval==0.2.2
pycparser==2.21
Pygments==2.14.0
pyparsing==3.0.9
pyrsistent==0.19.3
pyssim==0.7
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/python-dateutil_1626286286081/work
python-json-logger==2.0.7
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1673864280276/work
PyWavelets @ file:///home/conda/feedstock_root/build_artifacts/pywavelets_1673082327051/work
PyYAML @ file:///home/conda/feedstock_root/build_artifacts/pyyaml_1666772387118/work
pyzmq==25.0.2
requests==2.31.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
scikit-image @ file:///home/conda/feedstock_root/build_artifacts/scikit-image_1667117163409/work
scikit-learn @ file:///croot/scikit-learn_1690978916802/work
scikit-learn-intelex==20230426.111118
scipy @ file:///tmp/build/80754af9/scipy_1641555004408/work
Send2Trash==1.8.0
six @ file:///tmp/build/80754af9/six_1644875935023/work
sniffio==1.3.0
soupsieve==2.4
stack-data==0.6.2
statsmodels @ file:///home/conda/feedstock_root/build_artifacts/statsmodels_1667586005244/work
tabulate==0.9.0
tensorly @ file:///Users/jean/opt/anaconda3/conda-bld/tensorly_1673656889247/work
termcolor==2.3.0
terminado==0.17.1
thop==0.1.1.post2209072238
threadpoolctl @ file:///Users/ktietz/demo/mc3/conda-bld/threadpoolctl_1629802263681/work
tifffile @ file:///home/conda/feedstock_root/build_artifacts/tifffile_1617956128193/work
tinycss2==1.2.1
toolz @ file:///home/conda/feedstock_root/build_artifacts/toolz_1657485559105/work
torch==1.10.0
torchaudio==0.10.0
torchinfo @ file:///home/conda/feedstock_root/build_artifacts/torchinfo_1684097133203/work
torchmetrics @ file:///croot/torchmetrics_1687444060357/work
torchsummary==1.5.1
torchsummaryX==1.3.0
torchvision==0.11.0
tornado==6.2
tqdm==4.64.1
traitlets==5.9.0
typing_extensions @ file:///croot/typing_extensions_1669924550328/work
uri-template==1.2.0
urllib3==2.0.5
wcwidth==0.2.6
webcolors==1.13
webencodings==0.5.1
websocket-client==1.5.1
yacs==0.1.8
zipp==3.15.0
```
